fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -0,0 +1,1143 @@
+}
+
+void populateGetEncryptKeysByKeyIdsRequestBody(Reference<RESTKmsConnectorCtx> ctx,
+                                               const KmsConnLookupEKsByKeyIdsReq& req,
+                                               const bool refreshKmsUrls,
+                                               std::string& outJsonStr) {
+	// Acceptable request payload json format:
+	//
+	// request_json_payload {
+	//   "cipherKeyDetails" = [
+	//     {
+	//        "cipherBaseKeyId" : <cipherKeyId>
+	//        "encryptDomainId" : <domainId>
+	//     },
+	//     {
+	//         ....
+	//	   }
+	//   ],
+	//   "validationTokens" = [
+	//     {
+	//        "name" : <name>,
+	//        "value": <value>
+	//     },
+	//     {
+	//         ....
+	//     }
+	//   ]
+	//   "refreshKmsUrls" = 1/0
+	// }
+
+	rapidjson::Document doc;
+	doc.SetObject();
+
+	// Append "keyIdDetails" as json array
+	rapidjson::Value keyIdDetails(rapidjson::kArrayType);
+	for (const auto& detail : req.encryptKeyIds) {
+		rapidjson::Value keyIdDetail(rapidjson::kObjectType);
+
+		// Add "baseCipherId"
+		rapidjson::Value key(BASE_CIPHER_ID_TAG, doc.GetAllocator());
+		rapidjson::Value baseKeyId;
+		baseKeyId.SetUint64(detail.first);
+		keyIdDetail.AddMember(key, baseKeyId, doc.GetAllocator());
+
+		// Add "encryptDomainId"
+		key.SetString(ENCRYPT_DOMAIN_ID_TAG, doc.GetAllocator());
+		rapidjson::Value domainId;
+		domainId.SetInt64(detail.second);
+		keyIdDetail.AddMember(key, domainId, doc.GetAllocator());
+
+		// push above object to the array
+		keyIdDetails.PushBack(keyIdDetail, doc.GetAllocator());
+	}
+	rapidjson::Value memberKey(CIPHER_KEY_DETAILS_TAG, doc.GetAllocator());
+	doc.AddMember(memberKey, keyIdDetails, doc.GetAllocator());
+
+	// Append "validationTokens" as json array
+	addValidationTokensSectionToJsonDoc(ctx, doc);
+
+	// Append "refreshKmsUrls"
+	addRefreshKmsUrlsSectionToJsonDoc(ctx, doc, refreshKmsUrls);
+
+	// Serialize json to string
+	rapidjson::StringBuffer sb;
+	rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+	doc.Accept(writer);
+	outJsonStr.resize(sb.GetSize());
+	memcpy(outJsonStr.data(), sb.GetString(), sb.GetSize());
+}
+
+ACTOR Future<KmsConnLookupEKsByKeyIdsRep> fetchEncryptionKeyByKeyId(Reference<RESTKmsConnectorCtx> ctx,
+                                                                    KmsConnLookupEKsByKeyIdsReq req) {
+	state Reference<HTTP::Response> resp;
+	state KmsConnLookupEKsByKeyIdsRep reply;
+	state bool refreshKmsUrls = shouldRefreshKmsUrls(ctx);
+	state std::string requestBody;
+
+	populateGetEncryptKeysByKeyIdsRequestBody(ctx, req, refreshKmsUrls, requestBody);
+
+	// Follow 2-phase scheme:
+	// Phase-1: Attempt to fetch encryption keys by reaching out to cached KmsUrls in the order of
+	//          past success requests success counts.
+	// Phase-2: For some reason if none of the cached KmsUrls worked, re-discover the KmsUrls and
+	//          repeat phase-1.
+
+	state int pass = 1;
+	for (; pass <= 2; pass++) {
```

Comment:
Can we refactor all of the pass + retry logic into a common function, to abstract the general error handling and retry logic for contacting the Kms from the specific logic of each request type? Especially as we will be adding more connector rpc calls soon?

Reply:
Interesting thought, I can see code simplification for sure. Will take a crack at it. 
Thanks for pointing out.